### ISSUE TITLE
Encryption, Authentication, (Atmega32u4)

### ISSUE THREAD

**Issue by Timvrakas** [openLRSng/openLRSng#85]
Someone recently suggested that many drone application could benefit from encryption/authentication. Its a little beyond the scope of this project, but I thought it was a seed worth planting.

Would the AtMega328p be able to handle rudimentary crypto while doing the radio controls? How about a 32u4? If not, whats the next best thing? a co-processor? a 32-bit cpu? 2 x 32u4?

If you think its feasable, I may fork it and experiment.

Thanks!

**Comment by stasguru** [comment]
Hi KHA!
Is this feature still on your table?
Because it is really useful.

**Comment by kh4** [comment]
Yeah, this is somewhat on table. I am looking into integrating something like RC5 with a 64bit key. It should make 'attack' hard enough.

**Comment by frog32** [comment]
I do not see a huge benefit if we are not able to transmit the key to the receiver out of band. This could be achieved by hooking up the receiver to the ui.

**Comment by Timvrakas** [comment]
you could just use a simple RSA or DH protocol for the key exchange that
was linked to the binding process.
This might need to include an optional crypto chip system that could handle
the streamed data encryption.

Or you could just have the binding transmit the key at low power, and hope
no one is listening when you bind....

Sure makes an ARM Cortex look nice....

**Comment by kh4** [comment]
The plan is to use a random key transmitted to RX during binding. 

Yes it will be possible to grab the key during that, however when done properly the key will be over air just once. 

Notably this system does not need to be 'military grade' or anything, it is merely to make stealing control difficult enough.

**Comment by stipesb** [comment]
Have anyone success to put encryption in the openlrs? I want to make aes 256 encrypted communication with openlrs. Is it even possible?

**Comment by fiveangle** [comment]
There's no HW to support AES and the MCU isn't powerful enough, so not physically possible with current HW. Plenty of ways to obfuscate the link to prevent all but the most rigorous attempts (read above) but for AES, it would require different hardware

**Comment by mikeller** [comment]
@stipesb: Read https://github.com/openLRSng/openLRSng/issues/85#issuecomment-64344286

Not much point putting effort into adding something that will make the protocol illegal to use.

**Comment by stipesb** [comment]
@fiveangle  thanks for your answer and the info. You help me a lot.   

@mikeller I need this type of communication for some competition (they require aes 256). But modules for as 256 encrypted communication are too expensive. I didn't know that they are illegal to use, is it only in the US or Worldwide?

**Comment by mikeller** [comment]
@stipesb: It depends on what license you are planning to use openLRSng under - getting a commercial radio license will allow you to use encryption. But if you are using it under a ham radio license like most users do then encryption will be illegal - this is an ITU regulation and applicable world wide.

**Comment by stipesb** [comment]
@mikeller i would use it by commercial radio license. But if i succeed to make aes 256 encrypted communication i'm not allowed to share it for public?

**Comment by mikeller** [comment]
@stipesb: There is no restriction on sharing it, the restriction pertains to _using_ it for communication that is licensed under the ham radio regulations.
